✨ Feature/#28 rss 데이터 파싱 및 매핑

Atriel1999 opened the pull request:
## 📌 PR 유형 (해당하는 항목에 모두 체크해주세요)
- [x] Feat: 새로운 기능 추가
- [ ] Fix: 버그 수정
- [ ] Docs: 문서 수정
- [ ] Style: 코드 포맷팅, 세미콜론 누락, 코드 변경이 없는 경우
- [ ] Refactor: 코드 리팩토링 (기능 변경 없이 구조 개선)
- [ ] Test: 테스트 코드 추가 및 기존 테스트 리팩토링
- [x] Chore: 빌드 설정, 패키지 매니저 설정 등 기타 변경
- [ ] Github: PR 템플릿, 이슈 템플릿, Github Actions 설정 등
- [ ] Conflict: 머지 시 충돌 해결


## ✨ 변경 사항
- RSS 데이터 파싱 및 매핑 기능 구현
- 국민일보 RSS 피드를 파싱하여 공통 데이터 형식으로 매핑
- Quartz 스케줄러를 통한 주기적 데이터 수집 구현

## 🔍 리뷰어에게
- Mapper를 사용하여 다양한 뉴스 소스에 대응할 수 있도록 설계했습니다.
- RSS 소스 설정은 application-dev.yml에 있습니다. 디스코드 스레드에 별도로 공유하겠습니다.

TODO작업
     헥사고날 리팩토링
     수정된 뉴스 반영
     원격 MySQL로 application파일 수정 (현재는 local db에서 테스트함)

## ✅ PR 체크리스트
- [x] 커밋 메시지를 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항을 로컬에서 테스트했습니다.
- [x] 관련 라벨을 선택했습니다.


## 🔗 관련 이슈
- closes #28

Atriel1999 committed:
📦 Chore: application-dev.yml 파일 gitignore에 추가

Atriel1999 committed:
✨ Feat: RSS 뉴스 도메인 모델 및 기본 구조 구현

Atriel1999 committed:
✨ Feat: RSS Mapper 및 서비스 로직 구현

Atriel1999 committed:
✨ Feat: RSS 데이터 수집용 스케쥴러 Quartz 스케줄러 구현

chanbyoung reviewed (dismissed):
고생하셨어요!

엔터티 상위에 `@Data` 어노테이션을 사용하는 경우 id 필드도 setter가 생성되어 식별자 무결성이 깨질  수 있기 때문에 그 부분만 나중에 수정하면 될거 같아요!

YJ-circle commented on src/main/java/com/likelion/backendplus4/talkpick/batch/rss/service/RssService.java:
```diff
@@ -0,0 +1,134 @@
+        int savedCount = 0;
+
+        List<RssNews> newItems = newsItems.stream()
+                .filter(item -> !rssNewsRepository.existsByLink(item.getLink()))
```
코드 가독성 향상을 위해 이렇게 수정해보는건 어떨까요?

```suggestion
                .filter(not(item -> rssNewsRepository.existsByLink(item.getLink())))
```

아니면, 나중에 JPA 어댑터에서 notExist 메소드를 만들어주면 좋겠습니다..

Atriel1999 replied:
가독성으로 보면 첫번째 제안이 좋은거 같아서 반영하겠습니다
두번째 제안이신 notExist도 좋지만 헥사고날로 리팩토링하면서 아웃바운드에 그떄 추가하는게 편하지않을까 해서
일반 첫번째 제안을 반영하겠습니다.

thelightway24 reviewed (changes requested):
Mapper를 만드신 방식도 좋은데 공통 로직은 추상 클래스로 분리하고, 각 RSS마다 필요한 부분만 구현하는 구조가 유지보수나 확장 측면에서 훨씬 나을 것 같은데 어떻게 생각하시나요?

thelightway24 reviewed (commented):
구조를 바꾸는건 나중에 필요하면 진행하는 걸로 하면 좋겠습니다.

HaechangLee commented on .gitignore:
```diff
@@ -25,6 +25,7 @@ bin/
 out/
 !**/src/main/**/out/
 !**/src/test/**/out/
+application-dev.yml
```
application yml 파일은 github로 관리해도 괜찮다는 생각입니다.
지금 따로주신 application-dev.yml 에 DB정보가 들어가 있어서 ignore 걸어두신것 같은데, 나중엔 env로 빼고 수정해야하지 않을까요?

LEELISE replied:
네 저도 해창님과 같은 생각입니다

Atriel1999 replied:
env 반영했습니다!

Atriel1999 committed:
🐛Fix: RssNews 엔티티에서 @Data 어노테이션을 세분화된 Lombok 어노테이션으로 대체

- @Data 제거 후 @Getter로 전체 getter 생성
- 필요한 필드에만 선택적으로 @Setter 적용하여 식별자 무결성 보호
- @ToString에 description 필드 제외하여 로깅 최적화
- @EqualsAndHashCode에 id 필드만 사용하여 객체 비교 최적화

Atriel1999 commented on src/main/java/com/likelion/backendplus4/talkpick/batch/rss/config/RssSourceConfig.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.List;
+
+@Configuration
+@ConfigurationProperties(prefix = "rss")
```
현시점에선 application-dev.yml을 사용해서 거기에 등록해둔 rss 소스 정보를 가져오느라 사용됬습니다.
(현재는 외부설정 사용)

rss 소스 정보는 카테고리는 정해진대로 고정되있고 url도 자주변경되지않고 보안적으로도 상관없으니
아래 리뷰내용처럼 오히려 타입안전성을 보장해주는
enum으로 관리하는게 저도 좋아보여서 수정하는게 좋은거같습니다

Atriel1999 commented on src/main/java/com/likelion/backendplus4/talkpick/batch/rss/config/RssSourceConfig.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.ArrayList;
+import java.util.List;
+
+@Configuration
```
위 리뷰처럼 ConfigurationProperties를 사용해서 명시적으로 @Configuration이 더 좋다고 생각했습니다

Atriel1999 commented on src/main/java/com/likelion/backendplus4/talkpick/batch/rss/service/mapper/KmibRssMapper.java:
```diff
@@ -0,0 +1,78 @@
+        return link;
+    }
+
+    private String extractCategoryFromUrl(String url) {
+        if (url == null) return "";
+
+        if (url.contains("kmibPolRss")) return "정치";
+        if (url.contains("kmibEcoRss")) return "경제";
+        if (url.contains("kmibSocRss")) return "사회";
+        if (url.contains("kmibIntRss")) return "국제";
+        if (url.contains("kmibEntRss")) return "연예";
+        if (url.contains("kmibSpoRss")) return "스포츠";
```
아마 제가 로그를 부실하게 작성해서 발생한 문제같습니다.

해당메서드는 처음에 한번에 가져오는걸로 구현했다가
현재는 카테고리별로 가져오는걸로 수정했지만 (application-dev.yml에 지정되있음)
성능상에 문제가 없다고 생각해서 수정하지 않은 코드같습니다.

전체rss 스캔 -> 카테고리있음
카테고리별 rss 스캔 -> 이 역시 카테고리 데이터가있음 (정치 rss로 불러도 <category>안에 카테고리가있음)

중복적인 호출인거같아서 해당 메서드 대신 카테고리 정보를 직접사용하도록 수정하는거 좋을꺼같습니다.
ENUM으로 구조변경시 같이 진행하겠습니다

LEELISE commented on src/main/java/com/likelion/backendplus4/talkpick/batch/rss/scheduler/RssScheduler.java:
```diff
@@ -0,0 +1,29 @@
+
+    private final RssService rssService;
+
+    @Scheduled(cron = "0 */1 * * * ?")
```
P1:
운영과 관련된 핵심 설정들은 yml로 분리하는 것이 적절해보입니다!

Atriel1999 replied:
관련 설정을 application으로 분리하고 대신 값이 없을경우 기본값으로 1분이 되게 설정했습니다

LEELISE commented on src/main/java/com/likelion/backendplus4/talkpick/batch/rss/scheduler/RssScheduler.java:
```diff
@@ -0,0 +1,29 @@
+            int processedItems = rssService.fetchAndProcessAllFeeds();
+            log.info("Rss 피드 입력 Processed {} items", processedItems);
+        } catch (Exception e) {
+            log.error("Error during scheduled RSS feed fetch: {}", e.getMessage(), e);
```
P2:
알맞은 RssException 및 RssErrorCode가 필요해보입니다!

Atriel1999 replied:
exception과 errorcode를 만들어보고 서비스와 스케쥴러에 추가했습니다!

LEELISE commented on src/main/java/com/likelion/backendplus4/talkpick/batch/rss/service/RssService.java:
```diff
@@ -0,0 +1,134 @@
+                        List<RssNews> items = fetchFeedFromSource(source);
+                        allNewsItems.addAll(items);
+                    } catch (Exception e) {
+                        log.error("Error: 뉴스피드 fetch중 {}: {}", source.getName(), e.getMessage(), e);
```
P2:
알맞은 RssException 및 RssErrorCode가 필요해보입니다!

LEELISE commented on src/main/java/com/likelion/backendplus4/talkpick/batch/rss/service/mapper/KmibRssMapper.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    private LocalDateTime convertToLocalDateTime(Date date) {
+        return date != null
```
P5:
이미 정해져 있는 값과 정해져 있지 않은 값을 비교할 때 요다 컨디션에 따라 "null != date"의 식의 표현이 가독성 측면에서 더 권장될 수 있다고 합니다!

Atriel1999 replied:
요다 컨디션은 실수로 == 이렇게썻을때 에러도 방지할수도 있고
가독성 측면에서 좋은거같습니다.

반영했습니다 'v'b

Atriel1999 committed:
🐛 Fix: existsByLink 부정 조건을 not() 함수로 개선하여 가독성 향상

Atriel1999 committed:
✨ Feat: application-dev에서 prefix(카테고리)를 가져오는 구조에서 enum으로 변경

Atriel1999 committed:
📦 Chore: application-dev제거 및 env구조로 변경, env관련 의존성 추가(donenv)

Atriel1999 committed:
♻️ Refactor: prefix(언론사이름 접미사) 사용 구조 변경, 코드 구조 리팩토링

Atriel1999 committed:
♻️ Refactor: Quartz 실행주기 외부파일(application)으로 분리

Atriel1999 committed:
✨ Feat: 디버깅 관련 RssErrorCode & RssException 추가

Atriel1999 committed:
♻️ Refactor: 요다 조건식 변경

Atriel1999 committed:
Merge branch 'dev' into feature/#28-RSS-데이터-파싱-및-매핑

Atriel1999 commented:
@LEELISE 요청하신 모든 변경사항을 적용했습니다! 리뷰 상태를 업데이트해주실 수 있을까요?"

LEELISE commented:
> @LEELISE 요청하신 모든 변경사항을 적용했습니다! 리뷰 상태를 업데이트해주실 수 있을까요?"

고생하셨습니다!!!

LEELISE reviewed (approved):
와우 고생하셨습니다!!
